ocaml/libs/tracing/tracing.ml

```diff
@@ -0,0 +1,97 @@
+(*
+ * Copyright (C) 2023 Cloud Software Group
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+module SpanContext = struct
+  type t = {trace_id: string; span_id: string} [@@deriving rpcty]
+end
+
+module Span = struct
+  type t = {
+      span_context: SpanContext.t
+    ; span_parent: t option
+    ; span_name: string
+    ; span_begin_time: float
+    ; mutable span_end_time: float option
+    ; mutable tags: (string * string) list
+  }
+  [@@deriving rpcty]
+
+  let generate_id n = String.init n (fun _ -> "0123456789abcdef".[Random.int 16])
+
+  let start ?(tags = []) ~name ~parent () =
+    let trace_id =
+      match parent with
+      | None ->
+          generate_id 32
+      | Some span_parent ->
+          span_parent.span_context.trace_id
+    in
+    let span_id = generate_id 16 in
+    let span_context : SpanContext.t = {trace_id; span_id} in
+    let span_parent = parent in
+    let span_name = name in
+    let span_begin_time = Unix.gettimeofday () in
+    let span_end_time = None in
+    {span_context; span_parent; span_name; span_begin_time; span_end_time; tags}
+
+  let finish ?(tags = []) ~span () =
+    span.span_end_time <- Some (Unix.gettimeofday ()) ;
+    match (span.tags, tags) with
+    | _, [] ->
+        ()
+    | [], new_tags ->
+        span.tags <- new_tags
+    | orig_tags, new_tags ->
+        span.tags <- orig_tags @ new_tags
```

Comment:
Instead of using mutable fields, you can also construct a new value. Something like this:
```
{ span with
    span_end_time = Some (Unix.gettimeofday ()) ;
    tags = span.tags @ tags
}
```
But from this first patch it is not yet clear what the hash table is about. I guess your trying to update values in the hashtable, in place?

Reply:
The structure was remade to largely use less side effects, there is one instance where this cannot be done.